### ISSUE TITLE
3rd Party Cookie phase out - Firefox issue (Probably general issue)

### ISSUE THREAD

**Issue by deanmaster** [keycloak/keycloak#34812]
### Before reporting an issue

- [X] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

oidc

### Describe the bug

I had an email discussion with @jonkoops and he asked me to report the issue. 
I forked the repository to https://github.com/deanmaster/kjs-playground and playground 
I'm able to reproduce the problem with keycloak-js in the repository with my own Keycloak instance which runs in 25.0.x version. 
I assume this is also problem with latest Keycloak.

The main point which I changed from https://github.com/jonkoops/kjs-playground is I deploy keycloak in different domain and I don't use the same domain with different subdomain as it's written in the playground.

When I run directly from the fork. I can confirm these urls works without any problems:
```
- Keycloak Server - http://auth.keycloak.localhost:8080
- Keycloak Admin Console - http://admin.keycloak.localhost:8080
- Keycloak JS Playground - http://playground.keycloak.localhost:8080
```

The problem only occurs if you run Keycloak in different domain 
such as https://local.tuando.com (IDP) and http://playground.keycloak.localhost:8080(Application)

I explained my current solution in https://github.com/authts/oidc-client-ts/issues/1735 because i'm using oidc-client-ts library but not Keycloak adapter JS.

### Version

25.0.1

### Regression

- [x] The issue is a regression

### Expected behavior

Should be able to access to keycloak cookies.

### Actual behavior

As you can see in the How to Reproduce. You will see console from keycloak.js print out clearly that reading 3rd party cookie does not work.
Failed in step1.html

### How to Reproduce?

Deploy Keycloak in domain. https://local.tuando.com (my nginx domain in my local machine)
Create any client using OIDC, PKCE login flow.
Follow README in https://github.com/deanmaster/kjs-playground
Change main.ts to become
```
const keycloak = new Keycloak({
  url: 'https://local.tuando.com',
  realm: 'UAARealm',
  clientId: 'uaa-spa-client',
})
```
Start front-end only.
Access to http://playground.keycloak.localhost:8080
Observe the browser console
<img width="1908" alt="Screenshot 2024-11-11 at 08 15 56" src="https://github.com/user-attachments/assets/5a086107-88ab-4ff8-b72c-1bb37c1406f1">

### Anything else?

https://github.com/authts/oidc-client-ts/issues/1735 -> proposed solution for oidc-client-ts
https://github.com/keycloak/keycloak/issues/21307#issuecomment-2188994753 -> related issue to this

**Comment by deanmaster** [comment]
currently, our customer added Keycloak domain in exception in Firefox so the cookie can be shared. But still we should need to provide a better fix for that asking user changing their firefox configuration somehow still feel hacky.

**Comment by jonkoops** [comment]
I am going to need some steps to reproduce this in the latest version of Keycloak. The reproduction needs to be local, and not use any deployed instances of Keycloak in your own infrastructure, otherwise it will be very hard to fix the problem.

I've added [a commit](https://github.com/jonkoops/kjs-playground/commit/3d91802910f53c411de18c563ce58d8b08edac6f) to my playground application that will ensure that cross origin cookie behavior. This should be enough to reproduce the issue locally. If you manage to get this reproduced please also share any code modifications needed to trigger the behavior.

**Comment by deanmaster** [comment]
hello @jonkoops  thanks for having a look. I tried with your commit and I can see the error happen in my firefox locally. Token is expired after 60 seconds.

<img width="1914" alt="Screenshot 2024-11-25 at 05 01 43" src="https://github.com/user-attachments/assets/0e6fce42-6e82-433a-a2d0-6bc3a80c4984">

**Comment by jonkoops** [comment]
Can you please provide your changes either as a public commit on a repository or [code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks) in GitHub Markdown?

**Comment by deanmaster** [comment]
hi @jonkoops  I pushed everything to main in https://github.com/deanmaster/kjs-playground

**Comment by deanmaster** [comment]
hi @jonkoops let me setup a easy local domain with https and push the code.
I think using localhost as domain will not simulate correctly the issue, because it's still same top level domain.

I can also see using different domain rather than localhost will force to run with HTTPS because Native Webcrypto API.

**Comment by deanmaster** [comment]
i'm able to setup locally application run on different domain and I can see this Keycloak still be able to access cookie from IFrame. As we can see when it redirects back, the icone request 3rd party cookie is enable for domain "keycloak.localhost" But I also notice there is a warning in the console. I don't know what is that mean. 

Surprisingly, after direct back it's working. Event the storage access APIs really require user interaction such as Click on a Button. But I don't click on anything.

<img width="1906" alt="Screenshot 2024-11-26 at 07 18 55" src="https://github.com/user-attachments/assets/63d9f0cf-9472-4943-b62a-5e0a2fd8ebe8">

Next step, I will setup Keycloak locally on HTTPS in a easily way and commit the code.

**Comment by deanmaster** [comment]
I really don't understand how do we request unpartitioned cookies without any User Interaction ?

I can see there is a warning in the console and obviously the Keycloak (https://local.tmdo.com) has 3rd parties cookie access from Application (https://local.app.com:5173)

<img width="1916" alt="Screenshot 2024-11-26 at 08 21 22" src="https://github.com/user-attachments/assets/abdb4ffb-17b5-4a4a-921a-b1e4d00a4cca">

@jonkoops could you please explain to me how is our implementation ? This definitely can be implement in our general oidc-client-ts library which is a generic for OIDC.

**Comment by jonkoops** [comment]
> I think using localhost as domain will not simulate correctly the issue, because it's still same top level domain.

No, this is certainly a different eTLD+1. As `.localhost` is the eTLD, and `playground.localhost` is therefore different eTLD+1 than `*.keycloak.localhost`. You can clearly see the partitioning happening in the console, so it is honoring the heuristics that would be expected.

What could be the cause is that we might have a different browser configuration in what cookies are allowed. Yours could be stricter than mine.

> But I also notice there is a warning in the console. I don't know what is that mean.

I am not sure what might be causing this, but I suspect it is because you are mixing domains locally and on the web, on different ports, possibly because of self-signed certificates. Really, it could be for any reason, the dev tools never provide accurate information as to why you hit this heuristic, which is one of the most frustrating parts of debugging these kinds of issues.

As far as the heuristics go, you can't rely on them. I've tried as best as possible to work around them, but there are simply too many permutations if you add all the user preferences and browser inconsistencies together. The MDN documentation even [warns for this explicitly](https://developer.mozilla.org/en-US/docs/Web/Privacy/State_Partitioning#storage_access_heuristics):


> [!WARNING]  
> Storage access heuristics are a transitional feature meant to prevent website breakage. They should not be relied upon for current and future web development. 

> I think the best solution still giving a button for User to actively CLICK and allows 3rd parties access for 30 days (as far as I remember)

This is impossible, as this interaction needs to happen from a hidden iframe. This is how the OpenID Connect specification states it should be implemented, and we'd be deviating from the specification if we did this. The specification even goes as far as to acknowledge that this is essentially an [impossible problem to fix](https://openid.net/specs/openid-connect-session-1_0.html#ThirdPartyContent).

**Comment by deanmaster** [comment]
I totally agree with you @jonkoops From specification point of view there is no solution. 

that's why I propose/implemented the PoC in the abstract library oidc-client-ts. This need to be done either from libraries or from projects. 

From my PoC I have 2 separate IFrame:
* 1st IFrame is basically only show request button for click if I don't have access to Storage Access API. This IFrame will only display if I request and I don't have access permission for Storage otherwise IFrame would not shown. This IFrame has to serve in different endpoint but the same domain as IDP such as proxy together with Keycloak domain.
* 2nd IFrame is a place holder for future renew process happen, the placeholder url (which is another endpoint from IDP via proxy) will be replaced by property silent_renew.html url (build by library) with all necessary parameters. 

Keycloak or oidc-client-ts can only provide 2nd IFrame which you're doing right now. 

Long story short: The solution which i'm proposing and implemented is 

1. asking for permission from end-user specifically for Storage Access similar to asking for location or notification and it will show a specific popup for the user to click. 
2. Using Storage Access API normally for any new request. 

This will make sure we don't interrupt user for at least 30 days because the domain is trusted.

**Comment by jonkoops** [comment]
I don't think there is a resolution here for Keycloak itself, we're basically doing everything we can do considering the circumstances. If you want certain features implemented in a third-party client that can happen there, but you'd still have to load the iframe from the Keycloak server, and I don't think we can modify it any further without straying from the specification.
